src/framework_detectors.rs

```diff
@@ -0,0 +1,173 @@
+use regex::Regex;
+
+use crate::testrun::Framework;
+
+fn gen_reg(s: &str) -> Regex {
+    Regex::new(format!(r"(?i){}(\W|$)", s).as_str()).unwrap()
+}
+
+fn apply_reg(rl: &[(Regex, Framework)], v: Vec<String>) -> Option<Framework> {
+    for (r, f) in rl {
+        for s in v.iter() {
+            if r.is_match(s) {
+                return Some(*f);
+            }
+        }
+    }
+    None
+}
+
+fn get_framework_names() -> [(Regex, Framework); 4] {
+    [
+        (gen_reg("pytest"), Framework::Pytest),
+        (gen_reg("jest"), Framework::Jest),
+        (gen_reg("vitest"), Framework::Vitest),
+        (gen_reg("phpunit"), Framework::PHPUnit),
+    ]
+}
+
+fn get_file_extensions() -> [(Regex, Framework); 2] {
+    [
+        (gen_reg(".py"), Framework::Pytest),
+        (gen_reg(".php"), Framework::PHPUnit),
+    ]
+}
+
+// i want it to iterate through running certain regexes on all
+pub fn detect_framework(
+    testsuites_name: String,
+    mut testsuite_names: Vec<String>,
+    mut filenames: Vec<String>,
+    example_class_name: String,
+    example_test_name: String,
+    failure_messages: Vec<String>,
+) -> Option<Framework> {
+    let framework_names = get_framework_names();
+    testsuite_names.insert(0, testsuites_name);
+    match apply_reg(&framework_names, testsuite_names) {
```

Comment:
this could be done like this:

```rust
let names_to_match = std::iter::once(testsuites_name).chain(testsuite_names.iter());
for name in names_to_match {
  if let Some(framework) = find_matching_framework(name) {
    return Some(framework);
  }
}
```